Add support of rcar mmc driver

xakep-amatop opened the pull request:
Add basic functionality of Renesas SD/MMC driver. It can be used for both gen3 and gen4 R-car boards, but tested only with H3ULCB, Salvator XS M3 and Spider boards. This driver working with SDHC subsystem.

The driver supports regularal reading/writing throught SD/MMC controller buffer, DMA mode w/o interrupts.
Notes:
* the driver doesn't support SPI mode;
* SCC tuning and DMA mode based on IRQs are enabled by default;
* an address of a data buffer has to be aligned to 128 bytes if it is not, driver will use non-DMA mode automatically;
* Renesas MMC DMAC doesn't support 64-bit DMA addresses, so for case when we have 64-bit xref data address we use non-DMA mode;
* SD/MMC controller supports block size between 512 and 1 with a lot of restrictions, more details you can find in code;
* support of HS400 mode isn't implemented inside driver.

xakep-amatop committed:
drivers: sdhc: add support of Renesas MMC driver

Add basic functionality of Renesas SD/MMC driver. It can be used
for both gen3 and gen4 R-car boards, but tested only with H3ULCB,
Salvator XS M3 and Spider boards. This driver working with SDHC
subsystem.

The driver supports regularal reading/writing throught SD/MMC
controller buffer, DMA mode w/o interrupts and timing tuning.

Add gpio5 and sd0 nodes to h3ulcb and salvator xs which are needed
for working with SD cards. The GPIO node is needed for switching
voltage on SD card through gpio regulator driver.

Notes:
    * the driver doesn't support SPI mode;
    * SCC tuning and DMA mode based on IRQs are enabled by default;
    * an address of a data buffer has to be aligned to 128 bytes if it
      is not, driver will use non-DMA mode automatically;
    * Renesas MMC DMAC doesn't support 64-bit DMA addresses, so for
      case when we have 64-bit xref data address we use non-DMA mode;
    * SD/MMC controller supports block size between 512 and 1 with
      a lot of restrictions, more details you can find in code;
    * support of HS400 mode isn't implemented inside driver.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
boards: arm64: rcar: add board-specific configuration for MMC node

Add pin control group for UHS modes for H3ULCB, Salvator XS M3
boards to appropriate dts files. Both 'uhs' and 'default' pins
states have the same properties for eMMC, e.g. 1.8V.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
tests: add h3ulcb and salvator xs m3configs to perf disk tests

Add MMC related cfgs/overlays of arm64 gen3 rcar boards to perf and
susys/sd/mmc tests.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
samples: littlefs: add support for H3ULCB and Salvator XS M3

Add support of R-Car H3ULCB and Salvator XS M3 boards to
littlefs sample.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

danieldegrasse commented on drivers/regulator/regulator_fixed.c:
A few comments here:
- personally, I'd prefer this be in a second PR. That way the SD changes can be looked at independently of the regulator ones (and the regulator changes will pull in the appropriate reviewers)
- If we need these functions to be available to provide a common interface to regulators, could we implement them within `regulator_common.c`?
- ping @gmarull

xakep-amatop replied:
> personally, I'd prefer this be in a second PR. That way the SD changes can be looked at independently of the regulator ones (and the regulator changes will pull in the appropriate reviewers)

actually, GitHub CI will be red here due to lack of regulator changes, some of reviewers really worried about it

> If we need these functions to be available to provide a common interface to regulators, could we implement them within regulator_common.c?

I'm not sure about that. I think it's better to change this file. At least from now, it works in a similar way to the fixed regulator inside the Linux kernel. But we should consult with the maintainer of this driver, @gmarull, for the final decision.

danieldegrasse replied:
Based on your other comments in this PR, would the preferred regulator for this SD device be the driver being added in #58411?

xakep-amatop replied:
Yep, this pull request doesn't contain all the necessary changes for supporting SD cards. The driver is full, but we still need the following:
* pin control driver changes for controlling power on pins (https://github.com/zephyrproject-rtos/zephyr/pull/60444);
* GPIO driver changes (https://github.com/zephyrproject-rtos/zephyr/pull/58411) for controlling power of the SD card;
* also, some changes are needed for the board(s) DTS files (changes for our project repository can be found here: https://github.com/xen-troops/zephyr/pull/45).

The eMMC device on R-Car has only `fixed` regulators, and we can't disable power using these regulators, they are always on and fixed. If we talk about SD cards, they have both `fixed` and `gpio` regulators. One of them can disable the SD card power supply, and the other is used for controlling IO lines power.

jfischer-no replied:
Please copy this patch to own PR.

xakep-amatop replied:
ok, I'll do it

danieldegrasse commented on subsys/sd/sd.c:
```diff
@@ -127,7 +127,9 @@ static int sd_init_io(struct sd_card *card)
 	/* Cards start with legacy timing and 3.3V signalling at power on */
 	bus_io->timing = SDHC_TIMING_LEGACY;
-	bus_io->signal_voltage = SD_VOL_3_3_V;
+	bus_io->signal_voltage = card->host_props.host_caps.vol_330_support ?
```
Please update the comment here, and correctly record the signal voltage below: https://github.com/zephyrproject-rtos/zephyr/blob/18636af6d6658cf66e5f055901ffe2515116d329/subsys/sd/sd.c#L148

danieldegrasse commented on drivers/sdhc/Kconfig.rcar:
```diff
@@ -0,0 +1,41 @@
+	select REGULATOR
+	select GPIO
+	select SDHC_SUPPORTS_UHS if SDMMC_STACK
+	default n
```
This should be `default y`, with `depends on DT_HAS_RENESAS_RCAR_MMC_ENABLED`, see here: https://github.com/zephyrproject-rtos/zephyr/blob/18636af6d6658cf66e5f055901ffe2515116d329/drivers/sdhc/Kconfig.mcux_sdif#L7

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2388 @@
+		goto exit_unmap;
+	}
+
+	ret = rcar_mmc_init_start_clk(cfg);
```
In other SD drivers, the clock start is not run during device init. Rather, it is done within the `sdhc_request` implementation, typically based on CMD0 being sent. The reasoning here is that the clock init should be run after the SD card power is toggled, and the SD subsystem may call `sd_init_io` multiple times if the card fails init on the first attempt.

This is useful for cases such as a warm reset, where power may not have been toggled to the card, and the card may not be in an initialized state

xakep-amatop replied:
> In other SD drivers, the clock start is not run during device init.

Maybe I missed something, but let's take a look at these call chains:
```
imx_usdhc_init --> USDHC_Init --> CLOCK_EnableClock
```
and
```
mcux_sdif_init --> SDIF_Init --> CLOCK_EnableClock
``` 

In the case of my driver, this is the clock supply for the SD/MMC controller and the SD clock supply is controlled through the registers of the SD/MMC controller. As far as I understand, the same story applies to `nxp,lpc-sdif` and `nxp,imx-usdhc` as well.

-----
You are right, I leave the SD clock turned on at the end of the function  `rcar_mmc_init_controller_regs`.

Note, this controller supports the automatic disabling of SD clock signaling if there are no commands to SD/MMC. However, we have not enabled it yet, it requires some testing.

danieldegrasse replied:
Sorry, I am referring specifically to the SDMMC init sequence here (which is separate from clocking the SD host controller on all the controllers I have encountered). The SD spec requires sending 74 clock cycles to the SD card before CMD0 can be sent. This should either be done in `sdhc_set_io` when the clock is enabled, or in `sdhc_request` when CMD0 is sent.

danieldegrasse replied:
This only performs the delay. The spec itself requires that the controller send 74 clock cycles to the card before any commands can be sent. The USDHC and LPC SDIF controllers both also automatically disable clocks when not sending commands. However, they expose a register to trigger sending 74 clock cycles specifically for this portion of the specification:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/a12270da-5179-40c4-96c8-a338c9bc1a51)

If you do not have such a register on your hardware, you can also likely force the clock to be active for a millisecond or so to satisfy this requirement. Is there somewhere in this driver where this occurs that I am missing?

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2388 @@
+	uint32_t info1_reg;
+	int ret = 0;
+
+	if (dev_data->dma_support && ((uintptr_t)data->data % CONFIG_SDHC_BUFFER_ALIGNMENT == 0) &&
```
The SD subsystem should insure all data buffers passed to the driver are aligned to `CONFIG_SDHC_BUFFER_ALIGNMENT`, see here: https://github.com/zephyrproject-rtos/zephyr/blob/18636af6d6658cf66e5f055901ffe2515116d329/subsys/sd/sd_ops.c

So it should be possible to always use the DMA function when `dev_data->dma_support` is true unless I'm missing something specific to this controller?

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2388 @@
+#endif
+
+	host_caps->vol_330_support =
+		regulator_is_supported_voltage(cfg->regulator_vqmmc, 3300000, 3300000);
```
I'm confused what the regulator driver is used for here. It seems these fixed regulators cannot be enabled or disabled, and are just being used to determine the voltage ranges supported by the SDHC? Or is there a method to enable and disable these regulators I'm missing?

danieldegrasse commented on dts/arm64/renesas/rcar_gen3_ca57.dtsi:
```diff
@@ -28,6 +28,26 @@
+		regulator-name = "reg_3p3v";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-always-on;
```
I'm slightly confused about these regulator nodes:
- Should `always-on` be present, or `boot-on`? These regulators are not always enabled, and they are managed via the SDHC driver.
- Based on the changes to the fixed regulator driver and the lack of an enable pin, are these regulator nodes just being used to determine the voltage the SD host controller supports? If so, we should just make this a property of the host controller binding.

danieldegrasse replied:
Ok- I'm not familiar with the Rcar boards, but given that #58411 is merged, can this PR be updated to use that driver for either of these boards?

lorc replied:
Ah, I got this. This is for eMMC, when regulators are indeed fixed. It it strange that you didn't mentioned in the commit message, that you are adding DTS definitions for eMMC.

xakep-amatop replied:
Nothing seems strange to me. Feel free to walk through Zephyr commits in the 'dts' directory and look at the commit descriptions to see how other people have done it.

danieldegrasse commented on boards/arm64/rcar_salvator_xs_m3/rcar_salvator_xs_m3.dts:
```diff
@@ -30,3 +35,21 @@
+		     &emmc2_data0 &emmc2_data1 &emmc2_data2 &emmc2_data3
+		     &emmc2_data4 &emmc2_data5 &emmc2_data6 &emmc2_data7>;
+	pinctrl-names = "default", "uhs";
+	disk {
```
Why declare the disk here, if we are only going to enable it in overlays? Couldn't the disk be defined and enabled within the overlay itself?

danieldegrasse replied:
I would personally prefer we declare the disk in overlays, but I don't feel too strongly about it. Feel free to leave this as is

danieldegrasse reviewed (commented):
Driver itself generally looks good. Just a few comments/questions

Note- I'm not familiar with the hardware itself- I'd appreciate a review from anyone who is (@lorc)?

xakep-amatop commented:
Hi, @danieldegrasse @lorc @aaillet

I want to become a maintainer of this driver. Is it possible or not? Do I need to get additional approvals from someone else?

aaillet commented:
> Hi, @danieldegrasse @lorc @aaillet
> 
> I want to become a maintainer of this driver. Is it possible or not? Do I need to get additional approvals from someone else?

From my POV it's okay, add yourself as such in CODEOWNER file.
This file is supposed not to be used anymore but you will also become a collaborator after merge of #61066 so it's okay.
You can get more information about official "promotion" mechanism here : https://docs.zephyrproject.org/latest/project/project_roles.html

jfischer-no commented on drivers/sdhc/Kconfig.rcar:
```diff
@@ -0,0 +1,42 @@
+
+if RCAR_MMC_DMA_SUPPORT
+
+config SDHC_BUFFER_ALIGNMENT
+	default 128
```
Sounds like devicetree property.

danieldegrasse replied:
`SDHC_BUFFER_ALIGNMENT` is a Kconfig because it is intended to be raised from the default value by the SD host controller driver when needed. The idea for this Kconfig is that it can be used in order to force the SD subsystem to only provide aligned buffers to the SD host controller driver for data transfer. This can be helpful for operations like disk i/o, where we do not guarantee an alignment for buffers.

The reason this setting isn't defined in devicetree is because I view it as a software configuration property, not a hardware one. If we wanted to define SDHC alignment as a host controller property, we could change the SD subsystem to take that approach.

jfischer-no commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2388 @@
+#endif
+static int rcar_mmc_disable_scc(const struct device *dev);
+
+/**
+ * @brief Read 32-bit Renesas MMC register
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param dev MMC device, used for getting reg base address
+ * @param reg register offset relative to the base address
+ *
+ * @retval register value
+ */
```
What are the expectations of using Doxygen comments for driver-internal static functions?
Especially when you comment on something quite common and obvious?

xakep-amatop replied:
I aligned functions' descriptions with one another. As for me, well-documented code is welcomed in open communities. Moreover, a lot of drivers have Doxygen-style comments for driver-internal static functions.

jfischer-no replied:
> I aligned functions' descriptions with one another. As for me, well-documented code is welcomed in open communities.

`well-documented code !=  comment on something quite common and obvious`

>Moreover, a lot of drivers have Doxygen-style comments for driver-internal static functions.

No. And there is no reason to do same nonsense.

NACK.

xakep-amatop replied:
Is there any documented requirement that prohibits adding Doxygen-style comments for driver-internal static functions, or is it only your own preference?

jfischer-no replied:
No, but there is an undocumented preference not to do stupid things. Your driver is not a public API, so there is absolutely no reason to use Doxygen comments.

jfischer-no commented on drivers/sdhc/Kconfig.rcar:
```diff
@@ -0,0 +1,42 @@
+
+if MMC_RCAR
+
+config RCAR_MMC_DMA_SUPPORT
+	bool "Internal DMA support for Renesas Rcar MMC driver"
+	select CACHE_MANAGEMENT
+	select DCACHE
+	default y
+	help
+	  Internal DMA support for Renesas Rcar MMC driver.
+
+config RCAR_MMC_SCC_SUPPORT
+	bool "Support of SCC"
+	default y
+	help
+	  Enable support of Sampling Clock Controller for Renesas Rcar MMC driver.
```
Can it affect a particular instance of this driver? If so, use devicetree.

jfischer-no commented on drivers/sdhc/Kconfig.rcar:
```diff
@@ -0,0 +1,42 @@
+
+if MMC_RCAR
+
+config RCAR_MMC_DMA_SUPPORT
```
`
config MMC_RCAR_DMA_SUPPORT`

jfischer-no commented on tests/drivers/disk/disk_performance/boards/rcar_h3ulcb_ca57.conf:
```diff
@@ -0,0 +1,8 @@
+#
+
+CONFIG_SDHC=y
+CONFIG_MMC_RCA=1
```
Where is Kconfig option MMC_RCA introduced?

jfischer-no replied:
okay, with Kconfig options introduced here MMC_RCAR and RCAR_MMC_*, this MMC_RCA is very confusing.
Anyway, @danieldegrasse if a driver is designed to be the same for SD/MMC and eMMC controllers, then it looks like option MMC_RCA should not be Kconfig option.

danieldegrasse replied:
> okay, with Kconfig options introduced here MMC_RCAR and RCAR_MMC_*, this MMC_RCA is very confusing.
Anyway, @danieldegrasse if a driver is designed to be the same for SD/MMC and eMMC controllers, then it looks like option MMC_RCA should not be Kconfig option.

It is accurate that MMC_RCA should probably not be a Kconfig option. However, there is a bit more that would be required to support multiple cards on an MMC/SD bus. Once a card is in data transfer mode (after SET_RCA/CMD3 is sent), we could switch between active cards with CMD7 (GO_TRANSFER_STATE). This would require some additional changes to the subsystem, but I can implement it if I can find hardware that will enable testing multiple SD devices on one bus.

For the time being, I don't think this issue is in scope for this PR

jfischer-no commented on tests/subsys/sd/mmc/boards/rcar_h3ulcb_ca57.overlay:
```diff
@@ -0,0 +1,8 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+&emmc2 {
+	disk {
+		status = "okay";
+	};
+	status = "okay";
+};
```
Why is it not enabled on the board level, and why is it worthy to pollute samples and tests with these specific overlays?

jfischer-no replied:
> Because we don't need to enable everything that we support. Don't forget that this is a real-time OS, and some users may not want to use SD/MMC, for example. So, why should it be enabled by default?

So because of real-time, you say? okay, then do not enable the SD/MMC drivers by default, this is not usually done by node enabled status alone (but if it is, it is a bug).

jfischer-no replied:
> > this is not usually done by node enabled status alone (but if it is, it is a bug).
> 
> If I understand you correctly, your point is opposite to this one: [#60285 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/60285#discussion_r1262717048)

Usually there is major Kconfig option like SERIAL or here SDHC. Kconfig options like SERIAL or SDHC disable all drivers, regardless of node status and derived driver specific Kconfig option status. If emmc2 node describes hardware on your board then it should be described on the board level. But Kconfig option SDHC should not be enabled per default, see  https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#general-recommendations.

danieldegrasse replied:
What is being asked here is that the MMC device be added/enabled at the board level versus sample level. As @jfischer-no mentioned, unless `CONFIG_SDHC=y`, the SD host controller driver will not be enabled. You can see many example of this in Zephyr board devicetrees. For example, many boards enable SPI and I2C controllers, but those drivers won't be built unless `CONFIG_SPI` or `CONFIG_I2C` are turned on

danieldegrasse replied:
@xakep-amatop revisiting this- this PR has a *lot* of overlays, pretty much one per each test. Could you instead enable the common set of Kconfigs for these tests in the board level `Kconfig.defconfig`, like so?

```
if MMC_STACK

config MMC_RCA
	default 1

endif
```

gmarull reviewed (changes requested):
fixed regulators do not have any specific voltage by definition, also, accessing common driver class fields from drivers is wrong.

xakep-amatop commented:
> fixed regulators do not have any specific voltage by definition

In the Linux kernel, there are voltage fields for fixed regulators, and it is normal for other drivers to know the fixed range of voltage the regulator has. As per my understanding, 'fixed' means that the regulator is fixed at some voltage range and cannot be controlled to change this voltage range.

>  accessing common driver class fields from drivers is wrong

Sorry, I don't understand what you are trying to say here. Let me guess that you are talking about accessing common regulator fields from the fixed regulator driver. However, searching using only the word 'common' throughout the `drivers` directory gives a lot of matches:
```
gpio/gpio_andes_atcgpio100.c:297:	map &= dev_cfg->common.port_pin_mask;
gpio/gpio_eos_s3.c:367:	map &= config->common.port_pin_mask;
gpio/gpio_pca95xx.c:674:	if (BIT(pin) > config->common.port_pin_mask) {
gpio/gpio_xmc4xxx.c:90:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
gpio/gpio_xmc4xxx.c:149:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
gpio/gpio_xmc4xxx.c:170:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
gpio/gpio_xmc4xxx.c:183:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
gpio/gpio_xmc4xxx.c:202:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
gpio/gpio_nrfx.c:298:	map &= cfg->common.port_pin_mask;
gpio/gpio_emul.c:311:	if (~config->common.port_pin_mask & mask) {
gpio/gpio_emul.c:783:	map &= config->common.port_pin_mask;
gpio/gpio_xlnx_axi.c:92:	if (!(BIT(pin) & config->common.port_pin_mask)) {
gpio/gpio_mcux.c:276:	map &= config->common.port_pin_mask;
gpio/gpio_nct38xx_port.c:161:		if (mask & (~config->common.port_pin_mask)) {
gpio/gpio_nct38xx_port.c:472:		mask &= (enabled_gpios & config->common.port_pin_mask);
gpio/gpio_neorv32.c:62:	if (!(BIT(pin) & config->common.port_pin_mask)) {
gpio/gpio_sifive.c:297:	map &= cfg->common.port_pin_mask;

i3c/i3c_mcux.c:819:	return i3c_dev_list_find(&config->common.dev_list, id);
i3c/i3c_mcux.c:1237:			ret = i3c_dev_list_daa_addr_helper(&data->common.attached_dev.addr_slots,
i3c/i3c_mcux.c:1238:							   &config->common.dev_list, pid,
i3c/i3c_mcux.c:1251:			i3c_addr_slots_mark_i3c(&data->common.attached_dev.addr_slots, dyn_addr);

i3c/i3c_cdns.c:643:	struct i3c_config_controller *ctrl_config = &data->common.ctrl_config;
i3c/i3c_cdns.c:727:	if (!i3c_addr_slots_is_free(&data->common.attached_dev.addr_slots, controller_da)) {
i3c/i3c_cdns.c:729:			i3c_addr_slots_next_free_find(&data->common.attached_dev.addr_slots);
```

So, other drivers also access common class fields, of course, if I understand your point correctly.

gmarull commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -47,12 +51,49 @@ static int regulator_fixed_disable(const struct device *dev)
+
+static unsigned int regulator_fixed_count_voltages(const struct device *dev)
+{
+	return 1;
```
this is wrong for devices with no fixed voltage specified

gmarull reviewed (dismissed):
please see my comments, also open a PR specific to regulator changes to ease the review process.

gmarull commented:
> > fixed regulators do not have any specific voltage by definition
> 
> In the Linux kernel, there are voltage fields for fixed regulators, and it is normal for other drivers to know the fixed range of voltage the regulator has. As per my understanding, 'fixed' means that the regulator is fixed at some voltage range and cannot be controlled to change this voltage range.
> 
> Fair enough, I guess it's a particular case of min/max fields usage, because they are used to limit operating range in other cases.
> 
> > accessing common driver class fields from drivers is wrong
> 
> Sorry, I don't understand what you are trying to say here. Let me guess that you are talking about accessing common regulator fields from the fixed regulator driver. However, searching using only the word 'common' throughout the `drivers` directory gives a lot of matches:
> 
> ```
> gpio/gpio_andes_atcgpio100.c:297:	map &= dev_cfg->common.port_pin_mask;
> gpio/gpio_eos_s3.c:367:	map &= config->common.port_pin_mask;
> gpio/gpio_pca95xx.c:674:	if (BIT(pin) > config->common.port_pin_mask) {
> gpio/gpio_xmc4xxx.c:90:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
> gpio/gpio_xmc4xxx.c:149:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
> gpio/gpio_xmc4xxx.c:170:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
> gpio/gpio_xmc4xxx.c:183:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
> gpio/gpio_xmc4xxx.c:202:	gpio_port_pins_t pin_mask = config->common.port_pin_mask;
> gpio/gpio_nrfx.c:298:	map &= cfg->common.port_pin_mask;
> gpio/gpio_emul.c:311:	if (~config->common.port_pin_mask & mask) {
> gpio/gpio_emul.c:783:	map &= config->common.port_pin_mask;
> gpio/gpio_xlnx_axi.c:92:	if (!(BIT(pin) & config->common.port_pin_mask)) {
> gpio/gpio_mcux.c:276:	map &= config->common.port_pin_mask;
> gpio/gpio_nct38xx_port.c:161:		if (mask & (~config->common.port_pin_mask)) {
> gpio/gpio_nct38xx_port.c:472:		mask &= (enabled_gpios & config->common.port_pin_mask);
> gpio/gpio_neorv32.c:62:	if (!(BIT(pin) & config->common.port_pin_mask)) {
> gpio/gpio_sifive.c:297:	map &= cfg->common.port_pin_mask;
> 
> i3c/i3c_mcux.c:819:	return i3c_dev_list_find(&config->common.dev_list, id);
> i3c/i3c_mcux.c:1237:			ret = i3c_dev_list_daa_addr_helper(&data->common.attached_dev.addr_slots,
> i3c/i3c_mcux.c:1238:							   &config->common.dev_list, pid,
> i3c/i3c_mcux.c:1251:			i3c_addr_slots_mark_i3c(&data->common.attached_dev.addr_slots, dyn_addr);
> 
> i3c/i3c_cdns.c:643:	struct i3c_config_controller *ctrl_config = &data->common.ctrl_config;
> i3c/i3c_cdns.c:727:	if (!i3c_addr_slots_is_free(&data->common.attached_dev.addr_slots, controller_da)) {
> i3c/i3c_cdns.c:729:			i3c_addr_slots_next_free_find(&data->common.attached_dev.addr_slots);
> ```
> 
> So, other drivers also access common class fields, of course, if I understand your point correctly.
> 
> Regulator drivers class avoids doing this.

sorry, I edited your response, my mistake.

xakep-amatop commented:
> please see my comments, also open a PR specific to regulator changes to ease the review process.

I've opened a new PR for the regulator changes https://github.com/zephyrproject-rtos/zephyr/pull/61240 and fixed your comments there

xakep-amatop commented:
note: some of CI tests are red because of moving fixed regulator changes to another PR

xakep-amatop commented:
rebased without adding any new changes

danieldegrasse commented on dts/bindings/mmc/renesas,rcar-emmc.yaml:
```diff
@@ -0,0 +1,67 @@
+    default: 200000000
+    required: true
+
+  mmc-hs200-1_8v:
```
You shouldn't need these properties here- they are defined in `sdhc.yaml` already

danieldegrasse commented on dts/bindings/mmc/renesas,rcar-emmc.yaml:
```diff
@@ -0,0 +1,67 @@
+  pinctrl-names:
+    required: true
+
+  max-frequency:
```
Why not use `max-bus-freq` here?

xakep-amatop replied:
done

danieldegrasse commented on dts/bindings/mmc/renesas,rcar-emmc.yaml:
```diff
@@ -0,0 +1,67 @@
+  mmc-hs400-1_8v:
+    type: boolean
+
+  non-removable:
```
Please add a description to this property (although it is pretty self-explanatory, I still would like some text explaining that setting this property will affect `sdhc_card_present`)

danieldegrasse reviewed (commented):
Just a few comments on the binding. Otherwise looks good to go (once merge conflict is fixed)

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2252 @@
+	case SDHC_TIMING_LEGACY:
+		break;
+	case SDHC_TIMING_HS:
+		if (!data->props.host_caps.high_spd_support) {
```
These checks should not be needed. The SD stack will validate host support using `host_caps` fields before attempting to configure a given timing speed

xakep-amatop replied:
It isn't true if we're talking about the MMC stack, and this driver can be used for both SD and MMC cards. Look at the next call chain: `mmc_set_timing` → `mmc_set_hs_timing` (under the condition `ext->device_type.MMC_HS_52_DV`). The setting of the `MMC_HS_52_DV` field relies only on properties of the MMC card. For example, an MMC card can support this mode, but our controller might not support it.

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2252 @@
+
+	/* Set host driver type */
+	if (ios->driver_type && ios->driver_type != host_io->driver_type) {
+		switch (ios->driver_type) {
```
Why have this switch statement? It seems like the `sdhc_set_io` call will always return an error if the user tries to change the driver type

danieldegrasse replied:
Maybe I'm missing something here, sorry- if the user calls `sdhc_set_io` with a new driver type, wouldn't this switch statement fall through to the error? Or are you saying the SDMMC stack would never call `sdhc_set_io` with a changed driver type, since it will check the support fields?

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2252 @@
+	}
+#endif
+
+	if (unlikely(dev->data == NULL)) {
```
I don't think we need this check- how could the `dev->data` pointer be NULL given that it is initialized statically?

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2252 @@
+	struct mmc_rcar_data *data = dev->data;
+	const struct mmc_rcar_cfg *cfg = dev->config;
+
+	if (unlikely(cfg == NULL)) {
```
Same question here- I don't think that `dev->config` can be a NULL pointer given that it is set statically

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2252 @@
+	 * note: we need to set this flag before call rcar_mmc_init_controller_regs,
+	 *       because we want to cleanup dma regs and mask dma irqs
+	 */
+	data->dma_support = IS_ENABLED(CONFIG_RCAR_MMC_DMA_SUPPORT);
```
Why does this flag need to be in the `data` structure? It seems like we could enable DMA support in code by using `if(IS_ENABLED(CONFIG_RCAR_MMC_DMA_SUPPORT))`, or at a minimum declare `dma_support` as part of the device `config` struct so the definition is not stored in RAM

xakep-amatop replied:
There was a plan to add some shell functionality and the possibility to disable DMA during runtime.

I'll use define instead

xakep-amatop commented:
I rebased changes and fixed conflicts

danieldegrasse commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2252 @@
+	uint32_t info1_reg;
+	int ret = 0;
+
+	if (dev_data->dma_support && !(z_mem_phys_addr(data->data) >> 32)) {
```
The SD framework should check the alignment of the data buffer based on the setting of `SDHC_BUFFER_ALIGNMENT`, so I don't think this check is needed. Were you running into issues without it?

danieldegrasse commented on tests/subsys/sd/mmc/boards/rcar_h3ulcb_ca57.conf:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2023 EPAM Systems
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_DISK_ACCESS=y
```
Why do we need `CONFIG_DISK_ACCESS` set here?

danieldegrasse reviewed (commented):
@jfischer-no can you take another look at this PR, when you get a chance?

danieldegrasse reviewed (dismissed):
@jfischer-no can you revisit this PR? I'd like to move it forwards if possible

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -0,0 +1,2223 @@
+ * but we can run SDR104 on lower frequencies:
+ *    "SDR104: UHS-I 1.8V signaling, Frequency up to 208 MHz"
+ * so according to SD card standard it is possible to use lower frequencies,
+ * and we need to pass check of frequency in sdmmc in order to use sdr104 mode.
```
This looks a bit hacky for me. Could sdmmc core code be fixed to allow lower frequencies? I believe, this is the proper approach.

xakep-amatop replied:
I completely agree with you that it's a bit hacky. Would you mind if I open an issue for it, and then we can address it in a separate pull request?

lorc replied:
Actually yes, this is the great idea. You can link this discussion to the new issue.

xakep-amatop replied:
added an issue: https://github.com/zephyrproject-rtos/zephyr/issues/67943

lorc reviewed (dismissed):
> Note- I'm not familiar with the hardware itself- I'd appreciate a review from anyone who is (@lorc)?

Hardware-wise it is fine, I did extensive review internally  (it started at https://github.com/xen-troops/zephyr/pull/9 and then there were more pull-requests that shaped the driver in the final form).

danieldegrasse commented:
@jfischer-no Ping here- I think this driver is ready to go, and I'd like to get it moving forwards

jfischer-no reviewed (dismissed):
Problematic or unnecessary comment style not addressed. A large number of doxygen-style comments for no reason. Not only is this unnecessary and unused, but it bloats the code, requires maintenance, and causes unnecessary diff lines.
Also increases the driver's carbon footprint.

xakep-amatop commented:
> Problematic or unnecessary comment style not addressed. A large number of doxygen-style comments for no reason. Not only is this unnecessary and unused, but it bloats the code, requires maintenance, and causes unnecessary diff lines. Also increases the driver's carbon footprint.

Okay, could you please clarify what you need from me? I've cleaned up the code and deleted some of the Doxygen comments for small functions. Should I delete all the Doxygen comments for all functions, or should I convert them somehow? For example, before:
```
/**
 * @brief Check error flags inside INFO2 MMC register
 *
 * @note in/out parameters should be checked by a caller function
 *
 * @param dev MMC device
 *
 * @retval 0 INFO2 register hasn't errors
 * @retval -ETIMEDOUT: timed out while tx/rx
 * @retval -EIO: I/O error
 * @retval -EILSEQ: communication out of sync
 */
static int rcar_mmc_check_errors(const struct device *dev)
``` 
after
```
 /* check error flags inside INFO2 MMC register */
static int rcar_mmc_check_errors(const struct device *dev)
```

lorc commented:
Hi @jfischer-no,

May I ask you to place comments on offending pieces of PR, so we will be on the same page? 

> Also increases the driver's carbon footprint.

Running CI again and again after each pull request update increases carbon footprint significantly, because it requires much processing power to run all the checks.  So, if we really want to reduce carbon footprint we need to minimize number of the PR updates and for this we need to know exactly what should be fixed.

xakep-amatop commented:
fixed conflicts

danieldegrasse commented:
> fixed conflicts

Can you take a look at the failing testcases?

xakep-amatop commented:
> > fixed conflicts
> 
> Can you take a look at the failing testcases?

done

xakep-amatop commented:
@danieldegrasse @lorc please take another look at this pull request.

xakep-amatop commented:
fixed conflicts

danieldegrasse commented:
@jfischer-no can you please revisit this pull request?

xakep-amatop commented:
@danieldegrasse would you please review it again?

danieldegrasse commented:
> Problematic or unnecessary comment style not addressed. A large number of doxygen-style comments for no reason. Not only is this unnecessary and unused, but it bloats the code, requires maintenance, and causes unnecessary diff lines.
Also increases the driver's carbon footprint.

@jfischer-no is there a specific reason you don't want to use doxygen-style comments in the code for this PR? From what I can tell, this does not result in the functions being added to the Zephyr project API documentation, and other drivers in tree do use this style. If there is a code or style guideline I'm missing that would mandate this, please let me know. Otherwise, I do not want to block the movement of this PR based on this issue.

xakep-amatop commented:
@danieldegrasse ping

danieldegrasse commented:
> @jfischer-no is there a specific reason you don't want to use doxygen-style comments in the code for this PR? From what I can tell, this does not result in the functions being added to the Zephyr project API documentation, and other drivers in tree do use this style. If there is a code or style guideline I'm missing that would mandate this, please let me know. Otherwise, I do not want to continue to block the movement of this PR based on this issue.

@jfischer-no Please revisit your block (or add a comment clarifying your reasoning above) before Thursday  (April 22nd, 2024). Otherwise, following the criteria given by the project I will have to dismiss your review to in order to move this PR forwards.

(To be clear, the reason for dismissal would be [The reviewer did not revisit the pull request after 2 week and multiple pings by the author](https://docs.zephyrproject.org/latest/project/dev_env_and_tools.html#reviewers-should-keep-track-of-pull-requests-they-have-provided-feedback-to). While @xakep-amatop has not pinged you recently, I have done so multiple times on their behalf.)
